target_snowflake/sinks.py

```diff
@@ -1,44 +1,140 @@
+        )
+        self.connector.prepare_table(
+            self.full_table_name, schema=self.schema, primary_keys=self.key_properties
+        )
+        self.stage = self.stage_class(self)
 
     @property
     def max_size(self):
         return self.config["batch_size_rows"]
 
-    def start_batch(self, context: dict) -> None:
-        # TODO: perhaps Sync should have a callback hook at the beginning of execution?
-        if not self._has_migrated:
-            self.migrator.sync_table_schema()
-            self._has_migrated = True
-        super().start_batch(context)
+    def _write_csv(self, filepath: Path, records: Iterable[dict]) -> None:
+        """Write a CSV file."""
+        with open(filepath, "wt") as fp:
+            writer = csv.writer(fp, delimiter=",")
+            for i, record in enumerate(records, start=1):
+                if i == 1:
+                    writer.writerow(record.keys())
+
+                writer.writerow(record.values())
+        self.logger.info(f"Completed writing {i} records to file...")
+
+    def _upload_csv_to_table_stage(self, file_path: str, full_table_name: str) -> None:
+        """Upload data to the table's built-in stage."""
+        self.connector.connection.execute(
+            f"PUT file://{file_path} @%{full_table_name};"
+        )
+
+    def _load_csv_files_from_table_stage(self, full_table_name: str) -> None:
+        """Loads all files from the table's built-in stage."""
+        self.connector.connection.execute(
+            f"COPY INTO {full_table_name} "
+            f"FILE_FORMAT = (TYPE = CSV FIELD_DELIMITER = '\\t' SKIP_HEADER = 1);"
+        )
+
+    @property
+    def timestamp_time(self) -> datetime.datetime:
+        try:
+            return self._timestamp_time
+        except AttributeError:
+            self._timestamp_time = datetime.datetime.now(
+                tz=pytz.timezone(self.config["timestamp_timezone"])
+            )
+            return self._timestamp_time
+
+    @property
+    def filepath_replacement_map(self) -> Dict[str, str]:
+        return {
+            "stream_name": self.stream_name,
+            "datestamp": self.timestamp_time.strftime(self.config["datestamp_format"]),
+            "timestamp": self.timestamp_time.strftime(self.config["timestamp_format"]),
+        }
+
+    def _tmp_dir(self) -> Path:
+        """Return a temp directory path."""
+        return Path(tempfile.mkdtemp(prefix="target-snowflake"))
+
+    @property
+    def destination_path(self) -> Path:
+        result = str(self._tmp_dir / self.config["file_naming_scheme"])
+        for key, val in self.filepath_replacement_map.items():
+            replacement_pattern = "{" f"{key}" "}"
+            if replacement_pattern in result:
+                result = result.replace(replacement_pattern, val)
+
+        if self.config.get("output_path_prefix", None) is not None:
+            result = f"{self.config['output_path_prefix']}{result}"
```

Comment:
this is going to still have the tempfile path in between the prefix and filename, right? Is that a mistake?

Reply:
Very likely this was an oversight on my side. Feel free to fix/tweak.